### PR TITLE
support INFORMATION_SCHEMA

### PR DESCRIPTION
**Pull request by pgrivachev**
To support INFORMATION_SCHEMA we add corresponding catalogs if INFORMATION_SCHEMA exists in TableDefinition.

### CODE REVIEW

File: server/src/TableDefinition.ts [hunk 1]
```diff
@@ -14,20 +14,47 @@ export class TableDefinition {
...
+    return this.name[3].toLocaleLowerCase();
+  }
+
+  containsInformationSchema(): boolean {
+    return [this.name[1], this.name[2]].some(n => this.isInformationSchema(n));
+  }
+
+  isInformationSchema(namePart?: string): boolean {
+    return namePart?.toLocaleLowerCase() === 'information_schema';
+  }
```
Comment (fivetran-dmitryboykov) [inline server/src/TableDefinition.ts:57]: Do we really need to calculate it every time? May be we can make it once?

Comment (fivetran-dmitryboykov) [inline server/src/TableDefinition.ts:57]: Do we really need to calculate it every time? May be we can make it once?

File: server/src/TableDefinition.ts [hunk 2]
```diff
@@ -14,20 +14,47 @@ export class TableDefinition {
       this.name = name;
     }
 
-    this.datasetIndex = this.name.length === 3 ? 1 : 0;
+    this.datasetIndex = this.name.length >= 3 ? 1 : 0;
   }
 
   getProjectName(): string | undefined {
```
Comment (fivetran-dmitryboykov) [inline server/src/TableDefinition.ts:20]: Maybe we can calculate it once?

Comment (fivetran-dmitryboykov) [inline server/src/TableDefinition.ts:20]: Maybe we can calculate it once?

File: server/src/TableDefinition.ts [hunk 3]
```diff
@@ -14,20 +14,47 @@ export class TableDefinition {
...
+    if (this.containsInformationSchema()) {
+      if (this.isInformationSchema(this.name[1])) {
+        return undefined;
+      }
+      return this.name[0];
+    }
+    return this.name.length >= 3 ? this.name[0] : undefined;
   }
 
   getDatasetName(): string {
```
Comment (fivetran-dmitryboykov) [inline server/src/TableDefinition.ts:30]: Maybe we can calculate it once?

Comment (fivetran-dmitryboykov) [inline server/src/TableDefinition.ts:30]: Maybe we can calculate it once?

File: server/src/test/ZetaSqlWrapper.spec.ts [hunk 4]
```diff
@@ -35,55 +72,86 @@ describe('ZetaSqlWrapperTest', () => {
...
+    );
+  });
+
+  it('register should register information schema', async () => {
+    const tableName = 'columns';
+    const tableDefinition = new TableDefinition([PROJECT_ID, DATA_SET, InformationSchemaConfigurator.INFORMATION_SCHEMA, tableName]);
+    await shouldRegisterInformationSchema([tableDefinition], DATA_SET, tableName, PROJECT_ID);
+  });
+
+  it('register should register information schema when only schema specified', async () => {
```
Comment (fivetran-dmitryboykov) [inline server/src/test/ZetaSqlWrapper.spec.ts:151]: Do we need a test when only region specified? (like in e2e/projects/test-fixture/models/referrers.sql)

Comment (fivetran-dmitryboykov) [inline server/src/test/ZetaSqlWrapper.spec.ts:151]: Do we need a test when only region specified? (like in e2e/projects/test-fixture/models/referrers.sql)

**Review by fivetran-dmitryboykov — CHANGES_REQUESTED** [review]
Looks good. Just a couple of questions and suggestions.

File: server/src/InformationSchemaConfigurator.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
+import { SimpleCatalog, SimpleTable } from '@fivetrandevelopers/zetasql';
+import { TableDefinition } from './TableDefinition';
+import { ZetaSqlWrapper } from './ZetaSqlWrapper';
+
+export class InformationSchemaConfigurator {
+  static readonly INFORMATION_SCHEMA = 'information_schema';
+
+  static readonly INFORMATION_SCHEMA_COLUMNS = new Map<string, Map<string, string>>([
```
[the remark this answers is not in the document]
Comment (pgrivachev) [reply server/src/InformationSchemaConfigurator.ts:8]: Good point! Will add.

File: server/src/TableDefinition.ts (hunk 1 above)
Comment (pgrivachev) [reply server/src/TableDefinition.ts:57]: This method just compares any `namePart` to 'information_schema'

File: server/src/TableDefinition.ts [hunk 6]
```diff
@@ -14,20 +14,47 @@ export class TableDefinition {
       this.name = name;
     }
 
-    this.datasetIndex = this.name.length === 3 ? 1 : 0;
+    this.datasetIndex = this.name.length >= 3 ? 1 : 0;
```
[the remark this answers is not in the document]
Comment (pgrivachev) [reply server/src/TableDefinition.ts]: Yes, added tests.

File: server/src/test/ZetaSqlWrapper.spec.ts (hunk 4 above)
Comment (pgrivachev) [reply server/src/test/ZetaSqlWrapper.spec.ts:151]: I think it will be the same test as this in catalogs hierarchy perspective.

**Comment by fivetran-dmitryboykov** [comment]
`INFORMATION_SCHEMA` objects which not require dataset or region qualifier may cause errors for now. Let's try the next query:
```
select *
from INFORMATION_SCHEMA.SCHEMATA
```

**Comment by fivetran-dmitryboykov** [comment]
Can we catch somehow the next problem ?
```
select *
from NFORMATION_SCHEMA.INFORMATION_SCHEMA.JOBS_BY_PROJECT
```

**Review by fivetran-dmitryboykov — CHANGES_REQUESTED** [review]
Please address my comments.

**Comment by pgrivachev** [comment]
> select *
from NFORMATION_SCHEMA.INFORMATION_SCHEMA.JOBS_BY_PROJECT

Could you please clarify this case?

**Comment by fivetran-dmitryboykov** [comment]
> select *
> from NFORMATION_SCHEMA.INFORMATION_SCHEMA.JOBS_BY_PROJECT

Currently we consider this query as valid.

**Comment by pgrivachev** [comment]
> > select *
> > from NFORMATION_SCHEMA.INFORMATION_SCHEMA.JOBS_BY_PROJECT
> 
> Currently we consider this query as valid.
 
The problem here is that I don't know how to check if datasets like `region-us` exists. Api always shows that they are not exist. So I decided to consider all datasets/projectId as existing if name path contains `INFORMATION_SCHEMA`

**Pull request merged**